File: sharevox.ts
import bindings from "bindings";
const sharevoxCore: SharevoxCore = bindings("sharevox");

/**
 * sharevox_coreの利用を開始する
 * - 実行しないとsharevoxを利用できません
 * @param dlPath 動的ライブラリへのパス
 *
 */
export const initialize = (dlPath: string) =>
  new Promise<void>((resolve, reject) => {
    const { error } = sharevoxCore.initialize(String(dlPath));
    if (error) return reject(error);
    return resolve();
  });

/**
 * sharevox_coreの利用を終了する
 * - 実行するとsharevoxを利用ができなくなります
 * - 再度initializeを実行すると利用できます
 */
export const finalize = () =>
  new Promise<void>((resolve, reject) => {
    const { error } = sharevoxCore.finalize();
    if (error) return reject(error);
    return resolve();
  });

export const core = {
  /**
   * 初期化する
   * @param options 初期化オプション
   */
  sharevoxInitialize: (options: SharevoxInitializeOptions) =>
    new Promise<void>((resolve, reject) => {
      const { error } = sharevoxCore.sharevoxInitialize(
        options.accelerationMode,
        options.cpuNumThreads,
        options.loadAllModels,
        options.openJTalkDictDir
      );
      if (error) return reject(error);
      return resolve();
    }),
  /**
   * sharevoxのバージョンを取得する
   * @return SemVerでフォーマットされたバージョン
   */
  sharevoxGetVersion: () =>
    new Promise<string>((resolve, reject) => {
      const { error, result } = sharevoxCore.sharevoxGetVersion();
      if (error) return reject(error);
      return resolve(result);
    }),
  /**
   * モデルを読み込む
   * @param speakerId 読み込むモデルの話者ID
   */
  sharevoxLoadModel: (speakerId: number) =>
    new Promise<void>((resolve, reject) => {
      const { error } = sharevoxCore.sharevoxLoadModel(Number(speakerId));
      if (error) return reject(error);
      return resolve();
    }),
  /**
   * ハードウェアアクセラレーションがGPUモードか判定する
   * @return GPUモードならtrue、そうでないならfalse
   */
  sharevoxIsGpuMode: () =>
    new Promise<boolean>((resolve, reject) => {
      const { error, result } = sharevoxCore.sharevoxIsGpuMode();
      if (error) return reject(error);
      return resolve(result);
    }),
  /**
   * 指定したspeaker_idのモデルが読み込まれているか判定する
   * @return モデルが読み込まれているのであればtrue、そうでないならfalse
   */
  sharevoxIsModelLoaded: (speakerId: number) =>
    new Promise<boolean>((resolve, reject) => {
      const { error, result } = sharevoxCore.sharevoxIsModelLoaded(speakerId);
      if (error) return reject(error);
      return resolve(result);
    }),
  /**
   * このライブラリの利用を終了し、確保しているリソースを解放する
   */
  sharevoxFinalize: () =>
    new Promise<void>((resolve, reject) => {
      const { error } = sharevoxCore.sharevoxFinalize();
      if (error) return reject(error);
      return resolve();
    }),
  /**
   * メタ情報を取得する
   * @return メタ情報
   */
  sharevoxGetMetas: <T extends boolean>(getJsonString?: T) =>
    new Promise<T extends true ? string : Array<SharevoxMeta>>(
      (resolve, reject) => {
        const { error, result } = sharevoxCore.sharevoxGetMetasJson();
        if (error) return reject(error);
        return resolve(getJsonString ? result : JSON.parse(result));
      }
    ),
  /**
   * サポートデバイス情報を取得する
   * @return サポートデバイス情報
   */
  sharevoxGetSupportedDevices: <T extends boolean>(getJsonString?: T) =>
    new Promise<T extends true ? string : SharevoxSupportedDevices>(
      (resolve, reject) => {
        const { error, result } =
          sharevoxCore.sharevoxGetSupportedDevicesJson();
        if (error) return reject(error);
        return resolve(getJsonString ? result : JSON.parse(result));
      }
    ),
  /**
   * AudioQuery を実行する
   * @param text テキスト
   * @param speakerId 話者ID
   * @param options AudioQueryのオプション
   * @return AudioQuery
   */
  sharevoxAudioQuery: <T extends boolean>(
    text: string,
    speakerId: number,
    options: SharevoxAudioQueryOptions,
    getJsonString?: T
  ) =>
    new Promise<T extends true ? string : SharevoxAudioQuery>(
      (resolve, reject) => {
        const { error, result } = sharevoxCore.sharevoxAudioQuery(
          String(text),
          Number(speakerId),
          Boolean(options.kana)
        );
        if (error) return reject(error);
        return resolve(getJsonString ? result : JSON.parse(result));
      }
    ),
  /**
   * AudioQuery から音声合成する
   * @param audioQueryJson AudioQuery
   * @param speakerId  話者ID
   * @param options AudioQueryから音声合成オプション
   * @return wavデータ
   */
  sharevoxSynthesis: (
    audioQueryJson: string | SharevoxAudioQuery,
    speakerId: number,
    options: SharevoxSynthesisOptions
  ) =>
    new Promise<Buffer>((resolve, reject) => {
      const { error, result } = sharevoxCore.sharevoxSynthesis(
        typeof audioQueryJson === "object"
          ? JSON.stringify(audioQueryJson)
          : String(audioQueryJson),
        speakerId,
        options.enableInterrogativeUpspeak
      );
      if (error) return reject(error);
      return resolve(result);
    }),
};

interface SharevoxCore {
  initialize: (dlPath: string) => InitializeResult;
  finalize: () => FinalizeResult;
  sharevoxInitialize: (
    accelerationMode: number,
    cpuNumThreads: number,
    loadAllModels: boolean,
    openJTalkDictDir: string
  ) => SharevoxInitializeResult;
  sharevoxGetVersion: () => SharevoxGetVersionsResult;
  sharevoxLoadModel: (speakerId: number) => SharevoxLoadModelResult;
  sharevoxIsGpuMode: () => SharevoxIsGpuModeResult;
  sharevoxIsModelLoaded: (speakerId: number) => SharevoxIsModelLoadedResult;
  sharevoxFinalize: () => SharevoxFinalizeResult;
  sharevoxGetMetasJson: () => SharevoxGetMetasJsonResult;
  sharevoxGetSupportedDevicesJson: () => SharevoxGetSupportedDevicesResult;
  sharevoxAudioQuery: (
    text: string,
    speakerId: number,
    kana: boolean
  ) => SharevoxAudioQueryResult;
  sharevoxSynthesis: (
    audioQueryJson: string,
    speakerId: number,
    enableInterrogativeUpspeak: boolean
  ) => SharevoxSynthesisResult;
}
interface Result {
  error: string;
}
interface InitializeResult extends Result {}
interface FinalizeResult extends Result {}
interface SharevoxInitializeResult extends Result {}
interface SharevoxGetVersionsResult extends Result {
  result: string;
}
interface SharevoxLoadModelResult extends Result {}
interface SharevoxIsGpuModeResult extends Result {
  result: boolean;
}
interface SharevoxIsModelLoadedResult extends Result {
  result: boolean;
}
interface SharevoxFinalizeResult extends Result {}
interface SharevoxGetMetasJsonResult extends Result {
  result: string;
}
interface SharevoxGetSupportedDevicesResult extends Result {
  result: string;
}
interface SharevoxAudioQueryResult extends Result {
  result: string;
}
interface SharevoxSynthesisResult extends Result {
  result: Buffer;
}

/**
 * 初期化オプション
 */
export interface SharevoxInitializeOptions {
  /**
   * ハードウェアアクセラレーションモード
   */
  accelerationMode: SharevoxAccelerationMode;
  /**
   * CPU利用数を指定
   * 0を指定すると環境に合わせたCPUが利用される
   */
  cpuNumThreads: number;
  /**
   * 全てのモデルを読み込む
   */
  loadAllModels: boolean;
  /**
   * open_jtalkの辞書ディレクトリ
   */
  openJTalkDictDir: string;
}

/**
 * ハードウェアアクセラレーションモードを設定する設定値
 */
export enum SharevoxAccelerationMode {
  /**
   * 実行環境に合った適切なハードウェアアクセラレーションモードを選択する
   */
  SHAREVOX_ACCELERATION_MODE_AUTO = 0,
  /**
   * ハードウェアアクセラレーションモードを"CPU"に設定する
   */
  SHAREVOX_ACCELERATION_MODE_CPU = 1,
  /**
   * ハードウェアアクセラレーションモードを"GPU"に設定する
   */
  SHAREVOX_ACCELERATION_MODE_GPU = 2,
}

/**
 * Audio query のオプション
 */
export interface SharevoxAudioQueryOptions {
  /**
   * aquestalk形式のkanaとしてテキストを解釈する
   */
  kana: boolean;
}

/**
 * `sharevoxSynthesis` のオプション
 */
export interface SharevoxSynthesisOptions {
  /**
   * 疑問文の調整を有効にする
   */
  enableInterrogativeUpspeak: boolean;
}
/**
 * メタ情報
 */
interface SharevoxMeta {
  /**
   * 名前
   */
  name: string;
  /**
   * スピーカースタイルの一覧
   */
  styles: Array<SharevoxSpeakerStyle>;
  /**
   * スピーカーのUUID
   */
  speaker_uuid: string;
  /**
   * スピーカーのバージョン
   */
  version: string;
}
/**
 * スピーカースタイル
 */
interface SharevoxSpeakerStyle {
  /**
   * スタイル名
   */
  name: string;
  /**
   * スタイルID
   */
  id: number;
}
interface SharevoxSupportedDevices {
  /**
   * CPUに対応しているか
   */
  cpu: boolean;
  /**
   * CUDA(Nvidia GPU)に対応しているか
   */
  cuda: boolean;
  /**
   * DirectML(Nvidia GPU/Radeon GPU等)に対応しているか
   */
  dml: boolean;
}

/**
 * 音声合成クエリ
 */
interface SharevoxAudioQuery {
  /**
   * アクセント句のリスト
   */
  accent_phrases: Array<SharevoxAccentPhrase>;
  /**
   * 全体の話速
   */
  speed_scale: number;
  /**
   * 全体の音高
   */
  pitch_scale: number;
  /**
   * 全体の抑揚
   */
  intonation_scale: number;
  /**
   * 全体の音量
   */
  volume_scale: number;
  /**
   * 音声の前の無音時間
   */
  pre_phoneme_length: number;
  /**
   * 音声の後の無音時間
   */
  post_phoneme_length: number;
  /**
   * 音声データの出力サンプリングレート
   */
  output_sampling_rate: number;
  /**
   * 音声データをステレオ出力するか否か
   */
  output_stereo: boolean;
  /**
   * AquesTalkライクな読み仮名
   */
  kana: string;
}

/**
 * アクセント句
 */
interface SharevoxAccentPhrase {
  /**
   * モーラのリスト
   */
  moras: Array<SharevoxMora>;
  /**
   * アクセント箇所
   */
  accent: number;
  /**
   * 後ろに無音を付けるかどうか
   * モーラ（子音＋母音）ごとの情報
   */
  pause_mora: null | SharevoxMora;
  /**
   * 疑問系かどうか
   */
  is_interrogative: boolean;
}

/**
 * モーラ
 */
interface SharevoxMora {
  /**
   * 文字
   */
  text: string;
  /**
   * 子音の音素
   */
  consonant: null | string;
  /**
   * 子音の音長
   */
  consonant_length: null | number;
  /**
   * 母音の音素
   */
  vowel: string;
  /**
   * 母音の音長
   */
  vowel_length: number;
  /**
   * 音高
   */
  pitch: number;
}

File: sharevox.cc
#include <napi.h>
#include <string>
#include <dlfcn.h>
#include <iostream>


typedef struct 
{
  int32_t acceleration_mode = 0;
  uint16_t cpu_num_threads = 0;
  bool load_all_models = true;
  const char *open_jtalk_dict_dir = "";
} SharevoxInitializeOptions;

struct SharevoxAudioQueryOptions
{
  bool kana = false;
};
struct SharevoxSynthesisOptions
{
  bool enable_interrogative_upspeak = true;
};
void *sharevox_core = nullptr;
std::string load_string(const Napi::CallbackInfo &info, size_t index)
{
  return info[index].As<Napi::String>().Utf8Value() + "\0";
}
int32_t load_int32_t(const Napi::CallbackInfo &info, size_t index)
{
  return info[index].As<Napi::Number>().Int32Value();
}
uint32_t load_uint32_t(const Napi::CallbackInfo &info, size_t index)
{
  return info[index].As<Napi::Number>().Uint32Value();
}
bool load_bool(const Napi::CallbackInfo &info, size_t index)
{
  return info[index].As<Napi::Boolean>().Value();
}
template <typename T>
T load_func(const char *func_name, std::string &reason, bool &error)
{
  if (sharevox_core == NULL)
  {
    reason = std::string("sharevox_core not loaded.");
    error = true;
  }
  T func = reinterpret_cast<T>(NULL);
  if (!error)
    func = reinterpret_cast<T>(dlsym(sharevox_core, func_name));
  if (func == NULL)
  {
    reason = std::string("function \"") + std::string(func_name) + std::string("\" not found.");
    error = true;
  }
  return func;
}

std::string sharevox_core_error_result_to_message(int32_t &result)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<const char *(*)(int32_t)>("sharevox_error_result_to_message", reason, error);
  if (error)
  {
    return reason;
  }
  return std::string(func(result));
};

std::string sharevox_core_initialize(const char *root_dir_path, SharevoxInitializeOptions &options)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<int32_t (*)(const char *, SharevoxInitializeOptions)>("sharevox_initialize", reason, error);
  if (error)
  {
    return reason;
  }
  int32_t result = func(root_dir_path, options);
  if (result != 0)
  {
    return sharevox_core_error_result_to_message(result);
  }
  return std::string("");
};

std::string sharevox_core_get_version(std::string &version)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<const char *(*)(void)>("sharevox_get_version", reason, error);
  if (error)
  {
    return reason;
  }
  version = std::string(func());
  return std::string("");
};

std::string sharevox_core_load_model(uint32_t &speaker_id)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<int32_t (*)(uint32_t)>("sharevox_load_model", reason, error);
  if (error)
  {
    return reason;
  }
  int32_t result = func(speaker_id);
  if (result != 0)
  {
    return sharevox_core_error_result_to_message(result);
  }
  return std::string("");
};

std::string sharevox_core_is_gpu_mode(bool &is_gpu_mode)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<bool (*)(void)>("sharevox_is_gpu_mode", reason, error);
  if (error)
  {
    return reason;
  }
  is_gpu_mode = func();
  return std::string("");
};

std::string sharevox_core_is_model_loaded(uint32_t &speaker_id, bool &is_model_loaded)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<bool (*)(uint32_t)>("sharevox_is_model_loaded", reason, error);
  if (error)
  {
    return reason;
  }
  is_model_loaded = func(speaker_id);
  return std::string("");
};

std::string sharevox_core_finalize()
{
  std::string reason("");
  bool error = false;
  auto func = load_func<void (*)(void)>("sharevox_finalize", reason, error);
  if (error)
  {
    return reason;
  }
  func();
  return std::string("");
};

std::string sharevox_core_get_metas_json(std::string &metas_json)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<const char *(*)(void)>("sharevox_get_metas_json", reason, error);
  if (error)
  {
    return reason;
  }
  metas_json = std::string(func());
  return std::string("");
};

std::string sharevox_core_get_supported_devices_json(std::string &supported_devices_json)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<const char *(*)(void)>("sharevox_get_supported_devices_json", reason, error);
  if (error)
  {
    return reason;
  }
  supported_devices_json = std::string(func());
  return std::string("");
};

std::string sharevox_core_audio_query(const char *text, uint32_t &speaker_id, struct SharevoxAudioQueryOptions &options, char **output_audio_query_json)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<int32_t (*)(const char *, uint32_t, struct SharevoxAudioQueryOptions, char **)>("sharevox_audio_query", reason, error);
  if (error)
  {
    return reason;
  }
  int32_t result = func(text, speaker_id, options, output_audio_query_json);
  if (result != 0)
  {
    return sharevox_core_error_result_to_message(result);
  }
  return std::string("");
};

std::string sharevox_core_synthesis(const char *audio_query_json, uint32_t &speaker_id, struct SharevoxSynthesisOptions &options, uintptr_t *output_wav_length, uint8_t **output_wav)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<int32_t (*)(const char *, uint32_t, struct SharevoxSynthesisOptions, uintptr_t *, uint8_t **)>("sharevox_synthesis", reason, error);
  if (error)
  {
    return reason;
  }
  int32_t result = func(audio_query_json, speaker_id, options, output_wav_length, output_wav);
  if (result != 0)
  {
    return sharevox_core_error_result_to_message(result);
  }
  return std::string("");
};

std::string sharevox_core_audio_query_json_free(char *audio_query_json)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<void (*)(char *)>("sharevox_audio_query_json_free", reason, error);
  if (error)
  {
    return reason;
  }
  func(audio_query_json);
  return std::string("");
};

std::string sharevox_core_wav_free(uint8_t *wav)
{
  std::string reason("");
  bool error = false;
  auto func = load_func<void (*)(uint8_t *)>("sharevox_wav_free", reason, error);
  if (error)
  {
    return reason;
  }
  func(wav);
  return std::string("");
};

SharevoxInitializeOptions sharevox_core_make_default_initialize_options()
{
  std::string reason("");
  bool error = false;
  auto func = load_func<SharevoxInitializeOptions (*)(void)>("sharevox_make_default_initialize_options", reason, error);
  return func();
}

SharevoxAudioQueryOptions sharevox_core_make_default_audio_query_options()
{
  std::string reason("");
  bool error = false;
  auto func = load_func<SharevoxAudioQueryOptions (*)(void)>("sharevox_make_default_audio_query_options", reason, error);
  return func();
}

SharevoxSynthesisOptions sharevox_core_make_default_synthesis_options()
{
  std::string reason("");
  bool error = false;
  auto func = load_func<SharevoxSynthesisOptions (*)(void)>("sharevox_make_default_synthesis_options", reason, error);
  return func();
}

Napi::Object initialize(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  std::string dl_path = load_string(info, 0);
  sharevox_core = dlopen(dl_path.c_str(), RTLD_NOW);
  if (sharevox_core == NULL)
  {
    obj.Set("error", dlerror());
  }
  else
  {
    obj.Set("error", "");
  }
  return obj;
}

Napi::Object finalize(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  int result = dlclose(sharevox_core);
  obj.Set("result", "");
  if (result != 0)
  {
    obj.Set("error", dlerror());
  }
  else
  {
    obj.Set("error", "");
  }
  return obj;
}

Napi::Object sharevox_initialize(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  const char *root_dir_path = load_string(info, 4).c_str();
  std::string text = load_string(info, 3);
  char* cstr = new char[text.size() + 1];
  std::char_traits<char>::copy(cstr, text.c_str(), text.size() + 1);
  SharevoxInitializeOptions options = {
    load_int32_t(info, 0),
    static_cast<uint16_t>(load_uint32_t(info, 1)),
    load_bool(info, 2),
    const_cast<const char *>(cstr)
  };
  std::string error = sharevox_core_initialize(root_dir_path, options);
  obj.Set("error", error);
  return obj;
};
Napi::Object sharevox_get_version(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  std::string version("");
  std::string error = sharevox_core_get_version(version);
  obj.Set("error", error);
  obj.Set("result", version);
  return obj;
};

Napi::Object sharevox_load_model(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  uint32_t speaker_id = load_uint32_t(info, 0);
  std::string error = sharevox_core_load_model(speaker_id);
  obj.Set("error", error);
  return obj;
};

Napi::Object sharevox_is_gpu_mode(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  bool is_gpu_mode = false;
  std::string error = sharevox_core_is_gpu_mode(is_gpu_mode);
  obj.Set("error", error);
  obj.Set("result", is_gpu_mode);
  return obj;
};

Napi::Object sharevox_is_model_loaded(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  uint32_t speaker_id = load_uint32_t(info, 0);
  bool is_model_loaded = false;
  std::string error = sharevox_core_is_model_loaded(speaker_id, is_model_loaded);
  obj.Set("error", error);
  obj.Set("result", is_model_loaded);
  return obj;
};

Napi::Object sharevox_finalize(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  std::string error = sharevox_core_finalize();
  obj.Set("error", error);
  return obj;
};

Napi::Object sharevox_get_metas_json(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  std::string metas_json("");
  std::string error = sharevox_core_get_metas_json(metas_json);
  obj.Set("error", error);
  obj.Set("result", metas_json);
  return obj;
};

Napi::Object sharevox_get_supported_devices_json(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  std::string supported_devices("");
  std::string error = sharevox_core_get_supported_devices_json(supported_devices);
  obj.Set("error", error);
  obj.Set("result", supported_devices);
  return obj;
};

Napi::Object sharevox_audio_query(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  std::string text = load_string(info, 0);
  uint32_t speaker_id = load_uint32_t(info, 1);
  char *output_audio_query_json = nullptr;
  SharevoxAudioQueryOptions options = {false};
  options.kana = load_bool(info, 2);
  std::string error = sharevox_core_audio_query(text.c_str(), speaker_id, options, &output_audio_query_json);
  obj.Set("error", error);
  obj.Set("result", output_audio_query_json);
  std::string error_2 = sharevox_core_audio_query_json_free(output_audio_query_json);
  obj.Set("error_2", error_2);
  return obj;
};

Napi::Object sharevox_synthesis(const Napi::CallbackInfo &info)
{
  Napi::Env env = info.Env();
  Napi::Object obj = Napi::Object::New(env);
  std::string audio_query_json = load_string(info, 0);
  uint32_t speaker_id = load_uint32_t(info, 1);
  size_t output_wav_length = 0;
  uint8_t *output_wav = nullptr;
  SharevoxSynthesisOptions options = {false};
  options.enable_interrogative_upspeak = load_bool(info, 2);
  std::string error = sharevox_core_synthesis(audio_query_json.c_str(), speaker_id, options, &output_wav_length, &output_wav);
  obj.Set("error", error);
  obj.Set("result", Napi::Buffer<uint8_t>::New(env, output_wav, output_wav_length));
  std::string error_2 = sharevox_core_wav_free(output_wav);
  obj.Set("error_2", error_2);
  return obj;
};

Napi::Object Init(Napi::Env env, Napi::Object exports)
{
  exports.Set("initialize", Napi::Function::New(env, initialize));
  exports.Set("finalize", Napi::Function::New(env, finalize));

  exports.Set("sharevoxInitialize", Napi::Function::New(env, sharevox_initialize));
  exports.Set("sharevoxGetVersion", Napi::Function::New(env, sharevox_get_version));
  exports.Set("sharevoxLoadModel", Napi::Function::New(env, sharevox_load_model));
  exports.Set("sharevoxIsGpuMode", Napi::Function::New(env, sharevox_is_gpu_mode));
  exports.Set("sharevoxIsModelLoaded", Napi::Function::New(env, sharevox_is_model_loaded));
  exports.Set("sharevoxFinalize", Napi::Function::New(env, sharevox_finalize));
  exports.Set("sharevoxGetMetasJson", Napi::Function::New(env, sharevox_get_metas_json));
  exports.Set("sharevoxGetSupportedDevicesJson", Napi::Function::New(env, sharevox_get_supported_devices_json));
  exports.Set("sharevoxAudioQuery", Napi::Function::New(env, sharevox_audio_query));
  exports.Set("sharevoxSynthesis", Napi::Function::New(env, sharevox_synthesis));
  return exports;
}

NODE_API_MODULE(sharevox, Init);